File: examples/php/php-divergent_change-01_base/src/Controller/CourseStepsGetController.php
<?php

declare(strict_types=1);

namespace CodelyTv\DivergentChange\Controller;

use CodelyTv\DivergentChange\Platform;

final class CourseStepsGetController
{
    private Platform $platform;

    public function __construct(Platform $platform)
    {
        $this->platform = $platform;
    }

    public function get(string $courseId): string
    {
        $csv = $this->platform->findCourseSteps($courseId);
        if (empty($csv)) {
            return '[]';
        }

        $csvLines = explode(PHP_EOL, $csv);

        $steps = [];
        foreach ($csvLines as $row) {
            $row = str_getcsv($row);
            $type = $row[1];
            if ($type !== 'video' && $type !== 'quiz') {
                continue;
            }

            $durationInitialVideo = $row[3];
            $durationInitialQuiz = $row[2];
            $steps[] = [
                'id' => $row[0],
                'type' => $type,
                'duration' => $this->duration($type, $durationInitialVideo, $durationInitialQuiz),
                'points' => $this->points($type, $durationInitialVideo, $durationInitialQuiz)
            ];
        }

        $results = [];
        foreach ($steps as $step) {
            $results[] = json_encode($step, JSON_THROW_ON_ERROR);
        }

        $resultString = '[' . implode(',', $results) . ']';

        return $resultString;
    }

    private function duration(string $type, $durationVideo, $durationQuiz): float
    {
        $duration = 0;
        if ($type === 'video') {
            $duration = $durationVideo * 1.1; // 1.1 = due to video pauses
        }
        if ($type === 'quiz') {
            $duration = $durationQuiz * 0.5; // 0.5 = time in minutes per question
        }
        return $duration;
    }

    private function points(string $type, $durationVideo, $durationQuiz): float
    {
        $points = 0;
        if ($type === 'video') {
            $points = $durationVideo * 1.1 * 100;
        }
        if ($type === 'quiz') {
            $points = $durationQuiz * 0.5 * 10;
        }
        return $points;
    }
}

File: examples/csharp/csharp-courses_steps_csv-01_base/src/CoursesStepsCsv/CourseStepsGetController.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using Microsoft.VisualBasic.FileIO;

namespace CodelyTv.CoursesStepsCsv
{
    public sealed class CourseStepsGetController
    {
        private readonly Platform platform;

        public CourseStepsGetController(Platform platform)
        {
            this.platform = platform;
        }

        public string Get(string courseId)
        {
            var csv = platform.FindCourseSteps(courseId);

            var results = "[";

            var lines = csv.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries);

            for (int i = 0; i < lines.Length; i++)
            {
                var row = lines[i].Split(',');

                var type = row[1];
                var duration = 0.0;
                var points = 0.0;

                if (type == "video")
                {
                    duration = int.Parse(row[3]) * 1.1;
                }

                if (type == "quiz")
                {
                    duration = int.Parse(row[2]) * 0.5; // 0.5 = time in minutes per question
                }

                if (type != "video" && type != "quiz")
                {
                    continue;
                }

                if (type == "video")
                {
                    points = int.Parse(row[3]) * 1.1 * 100;
                }

                if (type == "quiz")
                {
                    points = int.Parse(row[2]) * 0.5 * 10;
                }

                var step = new Step
                {
                    Id = row[0],
                    Type = row[1],
                    Duration = duration,
                    Points = points
                };

                results += JsonSerializer.Serialize(step, new JsonSerializerOptions
                {
                    PropertyNamingPolicy = JsonNamingPolicy.CamelCase
                });

                if (i != lines.Length - 1)
                {
                    results += ",";
                }
            }
            results += "]";

            return results;
        }
    }
}
